// WizBulbAccess/lib/Scenes.js
'use strict';

const SCENESCONST = {
	OCEAN: 1,
	ROMANCE: 2,
	SUNSET: 3,
	PARTY: 4,
	FIREPLACE: 5,
	COZY: 6,
	FOREST: 7,
	PASTEL_COLORS: 8,
	WAKE_UP: 9,
	BEDTIME: 10,
	WARM_WHITE: 11,
	DAYLIGHT: 12,
	COOL_WHITE: 13,
	NIGHT_LIGHT: 14,
	FOCUS: 15,
	RELAX: 16,
	TRUE_COLORS: 17,
	TV_TIME: 18,
	PLANTGROWTH: 19,
	SPRING: 20,
	SUMMER: 21,
	FALL: 22,
	DEEPDIVE: 23,
	JUNGLE: 24,
	MOJITO: 25,
	CLUB: 26,
	CHRISTMAS: 27,
	HALLOWEEN: 28,
	CANDLELIGHT: 29,
	GOLDEN_WHITE: 30,
	PULSE: 31,
	STEAMPUNK: 32,
	RHYTHM: 1000,
}

class Scenes {

	var SCENES = ["Ocean","Romance","Sunset","Party","Fireplace","Cozy","Forest","Pastel Colors","Wake up","Bedtime","Warm White","Daylight",
	             "Cool white","Night light","Focus","Relax","True colors","TV time","Plantgrowth","Spring","Summer","Fall","Deepdive","Jungle",
	             "Mojito","Club","Christmas","Halloween","Candlelight","Golden white","Pulse","Steampunk","Rhythm" ];
	}

	getList() {
		return SCENES;
	}

	getCode(scenes) {
		var code = 1;
		var cont = 0;
		while(cont < 33) {
			if(scenes.localeCompare(SCENES[cont])) {
				cont = 50;
			} else {
				code++;
				cont++;
			}
		}
		return code;
	}
}


module.export = Scenes;
